src/Features/Core/Completion/AbstractCompletionService.cs

```diff
@@ -130,65 +135,146 @@ private class ProviderGroup
-            if (allGroups.Count == 0)
+            var allProvidersAndLists = nonExclusiveLists.Concat(nonUsedNonExclusiveProviders).ToList();
+            if (allProvidersAndLists.Count == 0)
             {
                 return null;
             }
 
             // Providers are ordered, but we processed them in our own order.  Ensure that the
             // groups are properly ordered based on the original providers.
-            allGroups.Sort((p1, p2) => completionProviderToIndex[p1.Provider] - completionProviderToIndex[p2.Provider]);
-            return allGroups.Select(g => g.Group);
+            allProvidersAndLists.Sort((p1, p2) => completionProviderToIndex[p1.Provider] - completionProviderToIndex[p2.Provider]);
+            return MergeAndPruneCompletionLists(allProvidersAndLists.Select(g => g.List), completionRules);
+        }
+
+        private static CompletionList MergeAndPruneCompletionLists(IEnumerable<CompletionList> completionLists, ICompletionRules completionRules)
+        {
+            var displayNameToItemsMap = new Dictionary<string, List<CompletionItem>>();
+            CompletionItem builder = null;
+
+            foreach (var completionList in completionLists)
+            {
+                if (completionList != null)
+                {
+                    foreach (var item in completionList.Items.WhereNotNull())
+                    {
+                        // New items that match an existing item will replace it.  
+                        ReplaceExistingItem(item, displayNameToItemsMap, completionRules);
+                    }
+
+                    builder = builder ?? completionList.Builder;
+                }
+            }
+
+            if (displayNameToItemsMap.Count == 0)
+            {
+                return null;
+            }
+
+            var totalItems = displayNameToItemsMap.Values.Flatten().ToList();
+            totalItems.Sort();
+
+            // TODO(DustinCa): This is lossy -- we lose the IsExclusive field. Fix that.
+
+            return new CompletionList(totalItems, builder);
         }
 
-        private Dictionary<ICompletionProvider, int> GetCompletionProviderToIndex(IEnumerable<ICompletionProvider> completionProviders)
+        private static void ReplaceExistingItem(
+            CompletionItem item,
+            Dictionary<string, List<CompletionItem>> displayNameToItemsMap,
+            ICompletionRules completionRules)
         {
-            var result = new Dictionary<ICompletionProvider, int>();
+            // See if we have an item with 
+            var sameNamedItems = displayNameToItemsMap.GetOrAdd(item.DisplayText, s_createList);
+            for (int i = 0; i < sameNamedItems.Count; i++)
+            {
+                var existingItem = sameNamedItems[i];
+
+                Contract.Assert(item.DisplayText == existingItem.DisplayText);
+
+                if (completionRules.ItemsMatch(item, existingItem).Value)
```

Comment:
Null won't be returned. I have a change a'comin' that makes all of the ICompletionRules just return bool rather than ```Nullable<bool>```.